google_guest_agent/snapshot_listener.go

```diff
@@ -0,0 +1,252 @@
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"strconv"
+	"syscall"
+	"time"
+
+	pb "github.com/GoogleCloudPlatform/guest-agent/google_guest_agent/protos"
+	"github.com/GoogleCloudPlatform/guest-logging-go/logger"
+	"github.com/hashicorp/golang-lru"
+	"google.golang.org/grpc"
+)
+
+const testPreSnapshotScript = "preSnapshotScript.sh"
+const testPostSnapshotScript = "postSnapshotScript.sh"
+const testIP = "169.254.0.1"
+const prodAddress = "169.254.169.254:8081"
+
+var (
+	seenPreSnapshotoperationIds, _  = lru.New(128)
+	seenPostSnapshotoperationIds, _ = lru.New(128)
+)
+
+type snapshotConfig struct {
+	Timeout               int // seconds
+	ContinueOnScriptError bool
+	PreSnapshotScriptURL  string
+	PostSnapshotScriptURL string
+	Enabled               bool
+}
+
+type scriptsTimeoutError struct {
+	path string
+}
+
+func (e *scriptsTimeoutError) Error() string {
+	return fmt.Sprintf("script timed out at: %s", e.path)
+}
+
+type invalidSnapshotConfig struct {
+	msg string
+}
+
+func (e *invalidSnapshotConfig) Error() string {
+	return fmt.Sprintf("invalid config: %s", e.msg)
+}
+
+func getSnapshotConfig() (snapshotConfig, error) {
+	setConfig()
+	var conf snapshotConfig
+	conf.Timeout = config.Section("Snapshots").Key("timeout_in_seconds").MustInt(60)
+	conf.ContinueOnScriptError = config.Section("Snapshots").Key("continue_on_script_error").MustBool(false)
+	conf.PreSnapshotScriptURL = config.Section("Snapshots").Key("pre_snapshot_script").MustString("")
+	conf.PostSnapshotScriptURL = config.Section("Snapshots").Key("post_snapshot_script").MustString("")
+
+	if conf.PreSnapshotScriptURL == "" && conf.PostSnapshotScriptURL == "" {
+		msg := "neither pre or post snapshot script has been configured"
+		logger.Errorf("%s", msg)
+		return conf, &invalidSnapshotConfig{msg}
+	}
+
+	return conf, nil
+}
+
+func runScript(path string, disks string, timeout int) (int, error) {
+	logger.Infof("running script at: %s", path)
+	cmd := exec.Command(path, disks)
+	err := cmd.Start()
+	if err != nil {
+		logger.Errorf("failed to start script at %s: %v", path, err)
+		return 0, err
+	}
+
+	done := make(chan error)
+	go func() { done <- cmd.Wait() }()
+
+	// Start timer
+	timer := time.After(time.Second * time.Duration(timeout))
+
+	select {
+	case <-timer:
+		logger.Errorf("timeout while running script at %s", path)
+		// Exit code of the timeout command when it times out
+		return 124, &scriptsTimeoutError{path}
+	case err := <-done:
+		if err != nil {
+			exiterr := err.(*exec.ExitError)
+			status := exiterr.Sys().(syscall.WaitStatus)
+			logger.Errorf("error while running script at %s, return with exit code %d", path, status.ExitStatus())
+			return status.ExitStatus(), nil
+		}
+	}
+	return 0, nil
+}
```

Comment:
this section should be simply a call to one of our existing runcmd functions. if there's a missing feature you need there (seems like timeout, maybe) then let's consider how we can extend.